src/client/src/producer.rs

```diff
@@ -96,82 +180,128 @@ impl TopicProducer {
         skip(self, buffer),
         fields(topic = %self.topic),
     )]
+    #[deprecated(since = "0.6.2", note = "Please use 'send_keyless' instead")]
     pub async fn send_record<B: AsRef<[u8]>>(
         &self,
         buffer: B,
-        partition: i32,
+        _partition: i32,
     ) -> Result<(), FluvioError> {
-        let record = buffer.as_ref();
-        let replica = ReplicaKey::new(&self.topic, partition);
-        debug!("sending records: {} bytes to: {}", record.len(), &replica);
+        let buffer: Vec<u8> = Vec::from(buffer.as_ref());
+        self.send_keyless(buffer).await?;
+        Ok(())
+    }
+}
 
-        let spu_client = self.pool.create_serial_socket(&replica).await?;
+/// A trait for defining a partitioning strategy for key/value records.
+///
+/// A Partitioner is given a slice of potential keys, and the number of
+/// partitions in the current Topic. It must map each key from the input
+/// slice into a partition stored at the same index in the output Vec.
+///
+/// It is up to the implementor to decide how the keys get mapped to
+/// partitions. This includes deciding what partition to assign to records
+/// with no keys (represented by `None` values in the keys slice).
+///
+/// See [`SiphashRoundRobinPartitioner`] for a reference implementation.
+trait Partitioner {
+    fn partition(&self, keys: &[Option<&[u8]>], partition_count: i32) -> Vec<i32>;
+}
 
-        debug!("connect to replica leader at: {}", spu_client);
+/// A [`Partitioner`] which combines hashing and round-robin partition assignment
+///
+/// - Records with keys get their keys hashed with siphash
+/// - Records without keys get assigned to partitions using round-robin
+struct SiphashRoundRobinPartitioner {
+    index: Arc<RwLock<i32>>,
```

Comment:
why this need to be in the lock?  A producer doesn't need to be thread safe nor shared

Reply:
> A producer doesn't need to be thread safe nor shared

Well, right now all of our producer's methods take self by immutable reference `&self`:

- `fn send<K, V>(&self, ...)`
- `fn send_keyless<V>(&self, ...)`
- `fn send_all<K, V>(&self, ...)`

And `TopicProducer` actually does implement Send and Sync. So in order to have mutable state we need to put the interior mutability somewhere. I am actually reworking some of this based on your earlier comment, let me ping you back on this when I have tweaked it a bit.